pkg/cli/upgrade/steps.go

```diff
@@ -0,0 +1,191 @@
+	crds := files.Filter("crds")
+	return u.kubeClient.ApplyManifestFile(crds, common.SystemNamespace)
+}
+
+func (u *Upgrade) upgradeHelmChart(ctx context.Context) error {
+	// We're already using the Helm chart, directly upgrade and return.
+	if u.helmReleaseExists {
+		u.l.Info("Upgrading Helm chart")
+		return u.helmInstaller.Upgrade(ctx)
+	}
+
+	// We're on the legacy installation.
+	// Perform migration of the existing resources to Helm.
+	u.l.Info("Migrating existing installation to Helm")
+	if err := u.cleanupLegacyResources(ctx); err != nil {
+		return fmt.Errorf("failed to cleanupLegacyResources: %w", err)
+	}
+	if err := u.helmInstaller.Install(ctx); err != nil {
+		return fmt.Errorf("failed to install Helm chart: %w", err)
+	}
+	dbNamespaces, err := u.kubeClient.GetDBNamespaces(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get database namespaces: %w", err)
+	}
+	for _, ns := range dbNamespaces {
+		if err := u.helmAdoptDBNamespaces(ctx, ns, u.upgradeToVersion); err != nil {
+			return fmt.Errorf("could not migrate DB namespace '%s' installation to Helm: %w", ns, err)
+		}
+	}
+	return nil
+}
+
+// Creates an installation of the `everest-db-namespace` Helm chart for the given DB namesapce
+// and adopts its resources.
+func (u *Upgrade) helmAdoptDBNamespaces(ctx context.Context, namespace, version string) error {
+	dbEngines, err := u.kubeClient.ListDatabaseEngines(ctx, namespace)
+	if err != nil {
+		return fmt.Errorf("cannot list database engines in namespace %s: %w", namespace, err)
+	}
+	values := helmutils.MustMergeValues(
+		helmValuesForDBEngines(dbEngines),
+		helm.ClusterValues(u.clusterType),
+	)
+	installer := helm.Installer{
+		ReleaseName:      namespace,
+		ReleaseNamespace: namespace,
+		Values:           values,
+	}
+	if err := installer.Init(u.config.KubeconfigPath, helm.ChartOptions{
+		URL:     u.config.RepoURL,
+		Name:    helm.EverestDBNamespaceChartName,
+		Version: version,
+	}); err != nil {
+		return fmt.Errorf("could not initialize Helm installer: %w", err)
+	}
+	return installer.Install(ctx)
+}
+
+func helmValuesForDBEngines(list *everestv1alpha1.DatabaseEngineList) values.Options {
+	vals := []string{}
+	for _, dbEngine := range list.Items {
+		t := dbEngine.Spec.Type
+		vals = append(vals, fmt.Sprintf("%s=%t", t, dbEngine.Status.State == everestv1alpha1.DBEngineStateInstalled))
+	}
+	// TODO: figure out how to set telemetry.
+	return values.Options{Values: vals}
+}
+
+// cleanupLegacyResources removes resources that were created as a part of the legacy installation method
+// and no longer exist as a part of the Helm chart.
+func (u *Upgrade) cleanupLegacyResources(ctx context.Context) error {
+	// Delete OLM PackageServer CSV.
+	if err := wait.PollUntilContextTimeout(ctx, pollInterval, pollTimeout, true, func(ctx context.Context) (bool, error) {
+		if err := u.kubeClient.DeleteClusterServiceVersion(ctx, types.NamespacedName{
+			Namespace: kubernetes.OLMNamespace,
+			Name:      "packageserver",
```

Comment:
Wouldn't this break some k8s clusters which have OLM preinstalled - i.e. openshift?

Reply:
This deletes the CRV only in `everest-olm` namespace, so I don't think it will break on Openshift since it has OLM in a different namespace